src/main/java/uk/gov/hmcts/darts/cache/token/service/AbstractTokenCache.java

```diff
@@ -299,6 +183,95 @@ private Optional<Token> getTokenValueWithResetExpiration(CacheValue holder) {
         return Optional.empty();
     }
 
+    private boolean getValidTokenFromCache(Token token, boolean validateToken) {
+        log.debug("Found the token in the cache");
+        if (validateToken && !token.validate(Token.TokenExpiryEnum.APPLY_EARLY_TOKEN_EXPIRY)) {
+            return false;
+        }
+
+        return true;
+    }
+
+    /**
+     * This function is key for storing to redis. It takes a service context and if reuse of token is required looks
+     * for the redis token and validates it (depending on the token representation). If the token is not valid then we aquire
+     * a new token and store in redis. If we do not require reuse a new token is generated each time.
+     * @param valueToBeCached The value to cache.
+     * @param reuseTokenIfPossible Do we want to use a shared token if we can.
+     * @param validateToken Do we wish to validate the token if one exists and sharing is required.
+     * @return The token to use. This maybe a shared token or a branch new one.
+     */
+    @SuppressWarnings("java:S6809")
+    private Optional<Token> createNewTokenOrReuseExistingToken(CacheValue valueToBeCached,
+                                                               Boolean reuseTokenIfPossible, boolean validateToken) throws CacheException {
+
+        log.debug("Storing the supplied value");
+
+        Optional<Token> tokenToUse = Optional.empty();
+
+        // if we have found a token, but it is invalid then invalidate it and force a new token to be created
+        if (reuseTokenBasedOnCredentials(reuseTokenIfPossible)) {
+            log.debug("Looking up token in cache");
+
+            // validate the token
+            tokenToUse = read(valueToBeCached);
+
+            if (tokenToUse.isPresent()) {
+                if (!getValidTokenFromCache(tokenToUse.get(), validateToken)) {
+                    tokenToUse = Optional.empty();
+                    log.debug("Cached token is invalid. Proceeding to create new token");
+                }
+            } else {
+                log.debug("Token not found in cache");
+            }
```

Comment:
in my previous comment, i meant we could put the whole of lines 217 -> 226 into the method `getValidTokenFromCache`